sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVector.java

```diff
@@ -126,12 +147,25 @@ public ArrayData copy() {
             list[i] = data.getLong(offset + i);
           }
         }
+      } else if (dt instanceof DecimalType) {
+        DecimalType decType = (DecimalType)dt;
+        for (int i = 0; i < length; i++) {
+          if (!data.getIsNull(offset + i)) {
+            list[i] = getDecimal(i, decType.precision(), decType.scale());
+          }
+        }
       } else if (dt instanceof StringType) {
         for (int i = 0; i < length; i++) {
           if (!data.getIsNull(offset + i)) {
             list[i] = ColumnVectorUtils.toString(data.getByteArray(offset + i));
           }
         }
+      } else if (dt instanceof CalendarIntervalType) {
```

Comment:
This is exposing missing implementation. We never generate arrays or complex types in the testing currently but implementing it is quite a bit more work. The issue is that to materialize the list of nested types inside the array, we need a deep copy of everything.

e.g. array<struct> means we need to copy the row for the struct. 

This patch is big enough that I'd prefer to defer. Also, this is extremely expensive so we might want to rethink how we do this. For example, having getArray return an iterator instead of the list.